internal/backends/winit/winitwindowadapter.rs

```diff
@@ -122,25 +125,44 @@ impl<Renderer: WinitCompatibleRenderer + 'static> WinitWindowAdapter<Renderer> {
     /// * `graphics_backend_factory`: The factor function stored in the GraphicsWindow that's called when the state
     ///   of the window changes to mapped. The event loop and window builder parameters can be used to create a
     ///   backing window.
-    pub(crate) fn new(#[cfg(target_arch = "wasm32")] canvas_id: String) -> Rc<dyn WindowAdapter> {
-        let self_rc = Rc::new_cyclic(|self_weak| Self {
-            window: corelib::api::Window::new(self_weak.clone() as _),
-            self_weak: self_weak.clone(),
-            map_state: RefCell::new(GraphicsWindowBackendState::Unmapped {
-                requested_position: None,
-                requested_size: None,
-            }),
-            currently_pressed_key_code: Default::default(),
-            pending_redraw: Cell::new(false),
-            in_resize_event: Cell::new(false),
-            dark_color_scheme: Default::default(),
-            renderer: Renderer::new(&(self_weak.clone() as _)),
-            #[cfg(target_arch = "wasm32")]
-            canvas_id,
-            #[cfg(target_arch = "wasm32")]
-            virtual_keyboard_helper: Default::default(),
+    pub(crate) fn new(
+        #[cfg(target_arch = "wasm32")] canvas_id: String,
+    ) -> Result<Rc<dyn WindowAdapter>, PlatformError> {
+        let self_rc = Rc::new_cyclic(|self_weak| {
+            let platform_error: Cell<Option<PlatformError>> = Default::default();
+            let mut renderer: once_cell::unsync::OnceCell<Renderer> = Default::default();
+
+            match Renderer::new(&(self_weak.clone() as _)) {
+                Ok(new_renderer) => {
+                    renderer = once_cell::unsync::OnceCell::with_value(new_renderer)
+                }
+                Err(err) => platform_error.set(Some(err)),
+            };
+
```

Comment:
This is annoying ugly.

Reply:
Yes, i'm afraid this is the only way for now.
https://github.com/rust-lang/libs-team/issues/90

One think that can a bit be improved would be to move the platform_error in the parent scope instead of in the Rc itself.
(See the code example in https://github.com/rust-lang/libs-team/issues/90#issuecomment-1304699842 )

Reply:
Ah yes, that would be nice. But then all the `WinitWindowAdapter` would have to support `Default`, i.e. `corelib::api::Window` have to go into a OnceCell. I'll give it a try to see how bad the code becomes - if too bad then I'll stick with this version.